File: 필수문항/170_인형들.py
import sys
import math

input = sys.stdin.readline


def get_b(i, k, m):
    global dolls
    s = 0
    for idx in range(i, i + k):
        s += (dolls[idx] - m) ** 2
    if s / k == 0.888888888888889:
        print("!!!")
    return s / k


n, k = map(int, input().split())
dolls = list(map(int, input().split()))
s = sum(dolls[:k])
ans = get_b(0, k, s / k)

tmp_sum = s

for i in range(1, n - k + 1):
    tmp_sum += dolls[k + i - 1]
    ans = min(ans, get_b(0, k + i, tmp_sum / (k + i)))


for now in range(1, n - k + 1):
    s -= dolls[now - 1] - dolls[now + k - 1]

    ans = min(ans, get_b(now, k, s / k))
    tmp_sum = s

    for i in range(1, n - k + 1 - now):
        tmp_sum += dolls[now + k + i - 1]
        ans = min(ans, get_b(now, k + i, tmp_sum / (k + i)))
print(math.sqrt(ans))

File: 필수문항/170_인형들.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Scanner;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int[] dolls = new int[n];
        for (int i = 0; i < n; i++) {
            dolls[i] = Integer.parseInt(st.nextToken());
        }

        double ans = solve(n, k, dolls);
        System.out.println(Math.sqrt(ans));
    }

    private static double getB(int[] dolls, int i, int k, double m) {
        double s = 0;
        for (int idx = i; idx < i + k; idx++) {
            s += Math.pow(dolls[idx] - m, 2);
        }
        return s / k;
    }

    private static double solve(int n, int k, int[] dolls) {
        double s = 0;
        for (int i = 0; i < k; i++) {
            s += dolls[i];
        }

        double ans = getB(dolls, 0, k, s / k);

        double tmpSum = s;

        for (int i = 1; i <= n - k; i++) {
            tmpSum += dolls[k + i - 1];
            ans = Math.min(ans, getB(dolls, 0, k + i, tmpSum / (k + i)));
        }

        for (int now = 1; now <= n - k; now++) {
            s -= dolls[now - 1] - dolls[now + k - 1];

            ans = Math.min(ans, getB(dolls, now, k, s / k));
            tmpSum = s;

            for (int i = 1; i <= n - k - now; i++) {
                tmpSum += dolls[now + k + i - 1];
                ans = Math.min(ans, getB(dolls, now, k + i, tmpSum / (k + i)));
            }
        }

        return ans;
    }
}
